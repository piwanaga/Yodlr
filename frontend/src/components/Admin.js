/** Admin page component
 * - shows all users
 * - can change user state from 'pending' to 'active' and vice versa
 * - can delete users
 */

import React, { useEffect, useState } from 'react';
import API from '../API';
import { Link } from 'react-router-dom';
import { Grid, makeStyles, Typography, Paper, Table, TableContainer, TableHead, TableRow, TableCell, TableBody, Button, IconButton } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';

const useStyles = makeStyles(() => ({
    header: {
        margin: '40px 0px'
    },
    dataGrid: {
        height: 600
    }
}));

const Admin = () => {
    const classes = useStyles();
    const [users, setUsers] = useState();
    const [userChanged, setUserChanged] = useState(true)

    /**Get list of users on component mount and when something is changed i.e. user is deleted or state is changed */
    useEffect(() => {
        const getUsers = async () => {
            const res = await API.getUsers();
            setUsers(res);
        };
        if (userChanged) {
            getUsers();
            setUserChanged(false)
        };
    }, [userChanged]);

    /**Call API to update user state, trigger effect via setUserChanged */
    const handleToggleState = async id => {
        await API.updateUserState(id);
        setUserChanged(true);
    };

    /**Call API to delete user, trigger effect via setUserChanged */
    const handleDelete = async id => {
        await API.deleteUser(id);
        setUserChanged(true);
    };

    return (
        <Grid container direction='column'>
            <Grid item>
                <Typography component={ Link } to='/'>Back to Home</Typography>
            </Grid>
            <Grid item>
                <Typography align='center' variant='h5' className={classes.header}>Users</Typography>
            </Grid>
            <Grid item className={classes.dataGrid}>
                {users ?
                <TableContainer component={ Paper }>
                    <Table>
                        <TableHead>
                            <TableRow>
                                <TableCell>Id</TableCell>
                                <TableCell>First Name</TableCell>
                                <TableCell>Last Name</TableCell>
                                <TableCell>Email</TableCell>
                                <TableCell>State</TableCell>
                                <TableCell>Toggle State</TableCell>
                                <TableCell>Delete User</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {users.map(user => (
                                <TableRow key={user.id}>
                                    <TableCell>{user.id}</TableCell>
                                    <TableCell>{user.firstName}</TableCell>
                                    <TableCell>{user.lastName}</TableCell>
                                    <TableCell>{user.email}</TableCell>
                                    <TableCell>{user.state}</TableCell>
                                    <TableCell>
                                        {user.state === 'active' ?
                                        <Button variant='outlined' color='secondary' onClick={() => handleToggleState(user.id)}>Deactivate</Button> :
                                        <Button variant='outlined' color='primary' onClick={() => handleToggleState(user.id)}>Set Active</Button>
                                        }
                                    </TableCell>
                                    <TableCell>
                                        <IconButton onClick={() => handleDelete(user.id)}>
                                            <DeleteIcon />
                                        </IconButton>
                                    </TableCell>
                                </TableRow>
                            ))}
                        </TableBody>
                    </Table>
                </TableContainer> :
                null
                }
            </Grid>
        </Grid>
    )
};

export default Admin;