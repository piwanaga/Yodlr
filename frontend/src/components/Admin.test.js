import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Admin from './Admin';


describe('Admin page showing all users', () => {
  let Wrapper

  beforeEach(() => {
    Wrapper = ({ children }) => (
      <BrowserRouter>
        {children}
      </BrowserRouter>
    );
  });

  it("passes snapshot test", () => {
    const { asFragment } = render(<Admin />, {wrapper: Wrapper});
    expect(asFragment()).toMatchSnapshot()
  });
});
