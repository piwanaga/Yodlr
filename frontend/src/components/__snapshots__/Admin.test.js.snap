// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin page showing all users passes snapshot test 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <a
        class="MuiTypography-root MuiTypography-body1"
        href="/"
      >
        Back to Home
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <h5
        class="MuiTypography-root makeStyles-header-1 MuiTypography-h5 MuiTypography-alignCenter"
      >
        Users
      </h5>
    </div>
    <div
      class="MuiGrid-root makeStyles-dataGrid-2 MuiGrid-item"
    />
  </div>
</DocumentFragment>
`;
